keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,981 @@
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+From [user_namespaces(7)](https://man7.org/linux/man-pages/man7/user_namespaces.7.html):
+> User namespaces isolate security-related identifiers and attributes, in
+particular, user IDs and group IDs, the root directory, keys, and capabilities.
+A process's user and group IDs can be different inside and outside a user
+namespace. In particular, a process can have a normal unprivileged user ID
+outside a user namespace while at the same time having a user ID of 0 inside
+the namespace; in other words, the process has full privileges for operations
+inside the user namespace, but is unprivileged for operations outside the
+namespace.
+
+The goal of supporting user namespaces in Kubernetes is to be able to run
+processes in pods with a different user and group IDs than in the host.
+Specifically, a privileged process in the pod runs as an unprivileged process in the
+host. If such a process is able to break into the host, it'll have limited
+impact as it'll be running as an unprivileged user there.
+
+The following security vulnerabilities are mitigated with
+user namespaces and it is expected that using them would mitigate against some of the future vulnerabilities.
+- CVE-2016-8867: Privilege escalation inside containers
+  - https://github.com/moby/moby/issues/27590
+- CVE-2018-15664: TOCTOU race attack that allows to read/write files in the host
+  - https://github.com/moby/moby/pull/39252
+- CVE-2019-5736: Host runc binary can be overwritten from container
+  - https://github.com/opencontainers/runc/commit/0a8e4117e7f715d5fbeef398405813ce8e88558b
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Increase node to pod isolation in Kubernetes by mapping user and group IDs
+  inside the container to different IDs in the host. In particular, mapping root
+  inside the container to unprivileged user and group IDs in the node.
+- Make it possible to run workloads that need "dangerous" capabilities such as
+  `CAP_SYS_ADMIN` without impacting the host.
+- Benefit from the security hardening that user namespaces are expected to
+  provide against some of the future unknown runtime vulnerabilities.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Provide a way to run the kubelet process or container runtimes as an
+  unprivileged process. Although initiatives like
+  [usernetes](https://github.com/rootless-containers/usernetes) and this KEP
+  both make use of user namespaces, it is a different implementation for a
+  different purpose.
+- Mounting volumes in pods with a user ID mapping. Although the authors of this
```

Comment:
Aren't we going to support volumes with cluster mappings?

Reply:
I think it's referring to supporting a shiftfs or similar feature. I clarified it in https://github.com/kinvolk/kubernetes-enhancements/pull/3/commits/1825e647568ae297c5cb0d5d85ee5541c7d57e43.
@alban could you please confirm if that's what you wanted to mention?

Reply:
Yes, that's what I meant. https://github.com/kinvolk/kubernetes-enhancements/commit/1825e647568ae297c5cb0d5d85ee5541c7d57e43 looks good to me